New interface

debris opened the pull request:
closes #118 
closes #119 

still todo: https://github.com/paritytech/ethabi/issues/119#issuecomment-416511101

cc @dvdplm @tomusdrw

tomusdrw commented on derive/src/function.rs:
```diff
@@ -0,0 +1,456 @@
+					}
+				}
+
+				pub struct Decoder(Function);
```
Would be good to add docs for the generated code.

tomusdrw commented on derive/src/function.rs:
```diff
@@ -0,0 +1,456 @@
+					ethabi::FunctionOutputDecoder::decode(&Decoder(Function::default()), output)
+				}
+
+				pub fn call<#(#declarations),*>(#(#definitions),*) -> (ethabi::Bytes, Decoder) {
```
I didn't see any higher-level tests for `call`, might be good to add some.

tomusdrw commented on derive/src/function.rs:
```diff
@@ -0,0 +1,456 @@
+				pub fn call<#(#declarations),*>(#(#definitions),*) -> (ethabi::Bytes, Decoder) {
+					let f = Function::default();
+					let tokens = vec![#(#tokenize),*];
+					(f.function.encode_input(&tokens).expect(INTERNAL_ERR), Decoder(f))
```
Why duplicate the code here? Maybe just call `encode_input` (or in future delegate to `Encoder` struct)

tomusdrw commented on derive/src/function.rs:
```diff
@@ -0,0 +1,456 @@
+				use ethabi;
+				use super::INTERNAL_ERR;
+
+				struct Function {
```
Why is this wrapper necessary/needed? Why not store `ethabi::Function` directly?

tomusdrw commented on derive/src/event.rs:
```diff
@@ -0,0 +1,399 @@
+					filter(#(#wildcard_filter_params),*)
+				}
+
+				pub fn parse_log(log: ethabi::RawLog) -> ethabi::Result<super::super::logs::#camel_name> {
```
Docs missing.

tomusdrw commented on derive/src/event.rs:
```diff
@@ -0,0 +1,399 @@
+					event: ethabi::Event,
+				}
+
+				impl Default for Event {
```
Again not really convinced we need a wrapper struct for this

tomusdrw commented on derive/src/contract.rs:
```diff
@@ -0,0 +1,88 @@
+
+			#constructor
+
+			pub mod functions {
```
docs

dvdplm commented on derive/src/constructor.rs:
```diff
@@ -0,0 +1,165 @@
+	to_ethabi_param_vec,
+};
+
+pub struct Constructor {
```
Some docs here would be good.

dvdplm commented on derive/src/event.rs:
```diff
@@ -0,0 +1,399 @@
+
+use super::{rust_type, to_syntax_string, from_token, get_template_names, to_token};
+
+pub struct Event {
```
Docs here

dvdplm commented on derive/src/event.rs:
```diff
@@ -0,0 +1,399 @@
+}
+
+impl Event {
+	pub fn generate_log(&self) -> quote::Tokens {
```
Docs.

debris commented:
> I like how readable the code and tests are. Still not sure how this is used from e.g. registrar in parity-ethereum – are you planning on changing the macro definition or make a follow-up PR to parity-ethereum?

Thanks @dvdplm . Yes, I do!

> Looks good, although the docs for generated code are missing and imho some of the structs are superfluous.

Thanks for the comments. I removed redundant wrappers and updated the docs for ethabi-derive structs. @tomusdrw do you also want to me to add docs for the generated code? I'm not convinced that this is necessary

dvdplm commented on derive/src/constructor.rs:
```diff
@@ -0,0 +1,131 @@
+	to_ethabi_param_vec,
+};
+
+/// Structure used to generate contract's construtor interface.
```
typo: "construtor"

tomusdrw commented:
@debris Yes, I think it would be good to document the generated code as well, otherwise it may trigger warnings if `#![warn(missing_docs)]` is enabled on the crate that uses the generator (and exposes some parts of it).

I think it will also help people understand what the functions are doing when browsing the documentation.
